protostar/starknet_gateway/gateway_facade.py

```diff
@@ -110,20 +117,36 @@ def _load_compiled_contract(self, compiled_contract_path: Path):
         except FileNotFoundError as err:
             raise CompilationOutputNotFoundException(compiled_contract_path) from err
 
-    def _transform_constructor_inputs_from_python(
+    def _transform_and_validate_constructor_inputs_from_python(
         self, inputs: Optional[CairoOrPythonData], compiled_contract_path: Path
     ) -> List[int]:
-        if not inputs:
-            return []
-        if not isinstance(inputs, collections.Mapping):
-            return inputs
         abi = self._compiled_contract_reader.load_abi_from_contract_path(
             compiled_contract_path
         )
         assert abi is not None
-        return from_python_transformer(abi, fn_name="constructor", mode="inputs")(
-            inputs
-        )
+
+        if not inputs:
+            # Convert empty dict and None into a list
+            inputs = []
+
+        if not has_abi_item(abi, "constructor"):
+            if inputs:
+                raise InputValidationException(
+                    "Inputs provided to a contract with no constructor."
+                )
+        else:
+            try:
+                if isinstance(inputs, collections.Mapping):
+                    inputs = from_python_transformer(
+                        abi, fn_name="constructor", mode="inputs"
+                    )(inputs)
+
+                # Validate inputs
+                to_python_transformer(abi, "constructor", "inputs")(inputs)
+            except DataTransformerException as ex:
+                raise InputValidationException(str(ex)) from ex
+
+        return cast(List[int], inputs)
```

Comment:
Can you add validation in a separate method?

Reply:
The problem is that `from_python_transformer` also performs some validation, so decoupling them is tricky.

Reply:
done